[APM] Related error count in transaction timeline

ogupte opened the pull request:
Closes #29688 by adding related error count in timeline with links to error overview filtered by the related trace/transaction ids, and only show the duration on hover. Also update the errors link in the sticky transaction properties to match the timeline pattern. 

![show-related-errors-in-timeline](https://user-images.githubusercontent.com/1967266/54714357-f5ae2700-4b0d-11e9-9c9f-483c9703f61d.gif)

ogupte committed:
Closes #29688 by adding related error count in timeline with links to error
overview filtered by the related trace/transaction ids. Also update the errors
link in the sticky transaction properties to match the timeline pattern.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5116/)

formgeist commented on x-pack/plugins/apm/public/style/variables.ts:
```diff
@@ -46,6 +46,10 @@ export const fontSizes = {
   xxlarge: '30px'
 };
 
+export const colors = {
```
Would prefer to import the EUI colours by `import theme from '@elastic/eui/dist/eui_theme_light.json';` and use the `euiColorDanger` from the theme. It's already been tweaked slightly in EUI since I made the designs, so the HEX doesn't match.

PS: Shed a small tear when you added the colors back in our variables because I went through and replaced them all with EUI a few weeks back 😢 I'd prefer going forward that we try and use their theming going forward. I'll make sure to reference them in my issues too 👍

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/WaterfallItem.tsx:
```diff
@@ -173,6 +176,12 @@ export function WaterfallItem({
         <HttpStatusCode item={item} />
         <NameLabel item={item} />
         <Duration item={item} />
+        {errorCount > 0 && item.docType === 'transaction' ? (
```
Can we move this up before `Duration`? My intention was to also only show `Duration` on row hover, but that probably wasn't clear from the issue description 😬We can implement that in another PR, if we don't want to mess with this right now.

![image](https://user-images.githubusercontent.com/4104278/54268673-0db90000-457c-11e9-928b-2e6820bf34dd.png)

formgeist replied:
Ah, that looks great... when all the labels are aligned left. When the spans are further on the right, and the label information aligns on the right, the duration should show on the left, because it looks a little strange when we reserve the space for the duration to be shown. Not sure how easily achievable this is? I've tried to figure out how we align the labels, especially when the span is on on the right.

<img width="719" alt="Screenshot 2019-03-14 at 09 06 03" src="https://user-images.githubusercontent.com/4104278/54340974-ed02b000-4638-11e9-8590-8aa90d4bf0d7.png">

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/StickyTransactionProperties.tsx:
```diff
@@ -109,16 +88,7 @@ export function StickyTransactionProperties({
-      val: idx(transaction, _ => _.user.id) || NOT_AVAILABLE_LABEL,
-      truncated: true,
-      width: '25%'
+      width: '10%'
```
Sorry, I tested it out locally and felt that 10% was probably too little space on smaller screens for the result to render on one line (example in your screenshot in the description). Can we change this to `14%`?

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/StickyTransactionProperties.tsx:
```diff
@@ -127,8 +97,25 @@ export function StickyTransactionProperties({
+      ) : (
+        noErrorsText
+      ),
+      width: '20%'
```
Should be updated to `18%` because we're increasing the Result item.

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/StickyTransactionProperties.tsx:
```diff
@@ -127,8 +97,25 @@ export function StickyTransactionProperties({
+      fieldName: USER_ID,
+      val: idx(transaction, _ => _.user.id) || NOT_AVAILABLE_LABEL,
+      truncated: true,
+      width: '20%'
```
Should be updated to `18%` because we're increasing the Result item.

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,82 @@
+
+const LinkLabel = styled.span`
+  font-size: ${fontSizes.small};
+  color: ${colors.danger};
```
Should be updated to use EUI colors (see my comment about EUI theme colours).

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,82 @@
+import { colors, fontSizes } from '../../../../style/variables';
+
+const LinkLabel = styled.span`
+  font-size: ${fontSizes.small};
```
It's a little on the small side, we should use `fontSize` instead so it renders as 14px.

formgeist reviewed (changes requested):
@ogupte Thanks taking this on so quick 👍 Made some comments mostly around how we use colors through EUI theming and some minor design tweaks.

ogupte committed:
[APM] leverage EUI theme colors, tweak sticky transaction properties column
sizes, and reorder timeline error count to show duration after only on hover

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5338/)

formgeist commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/StickyTransactionProperties.tsx:
```diff
@@ -127,8 +97,25 @@ export function StickyTransactionProperties({
-      val: errorCount ? errorsOverviewLink : noErrorsText,
-      width: '25%'
+      val: errorCount ? (
+        <ErrorsOverviewLink
```
Would it be possible to disable the tooltip showing "View 2 related errors" in this place, as I think the suffix label "Related errors" makes the tooltip superfluous?

<img width="502" alt="Screenshot 2019-03-14 at 09 05 54" src="https://user-images.githubusercontent.com/4104278/54341097-4965cf80-4639-11e9-8047-3ef03c6db60a.png">

sqren replied:
@ogupte Maybe the component should have a `verbose` options. If `true` it show the suffix label, if false it has the tooltip.

ogupte replied:
> Ah, that looks great... when all the labels are aligned left. When the spans are further on the right, and the label information aligns on the right, the duration should show on the left, because it looks a little strange when we reserve the space for the duration to be shown. Not sure how easily achievable this is? I've tried to figure out how we align the labels, especially when the span is on on the right.

I think we should create a new ticket for this since there's some non-trivial to determine the alignment within the waterfall.

formgeist replied:
@ogupte OK, let's move the show/hide functionality out of this PR too, I think it's too distracting as is, and not critical to implement the related errors count on the transaction span row. Perhaps we can open a PR with code, but have a task that has to look at the directional placement. What do you think?

formgeist reviewed (changes requested):
@ogupte Thanks for the updates, looking really good. I found some issues that I should have declared earlier.

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,84 @@
+      <EuiToolTip content={toolTipContent}>
+        <span>
+          <EuiBadge
+            color={euiThemeLight.euiColorDanger}
```
Unless we need a custom color, we can just use a string: `<EuiBadge color="danger">`
@formgeist Do you have any preference?

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,84 @@
+}
+
+export const ErrorsOverviewLink: React.SFC<ErrorsOverviewLinkProps> = ({
+  errorCount,
```
Perhaps set a default `errorCount = 0`, then you can remove the defaults in the component.

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,84 @@
+          <EuiBadge
+            color={euiThemeLight.euiColorDanger}
+            onClick={(event: any) => {
+              (event as MouseEvent).stopPropagation();
```
Are the EUI types incorrect? If I change `EuiBadge` to a `div` the type of `event` is inferred.

ogupte replied:
i the type wasn't inferred for me when using the EuiBadge component, could be a missing type there.

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/ErrorsOverviewLink.tsx:
```diff
@@ -0,0 +1,84 @@
+  showLabel?: boolean;
+}
+
+export const ErrorsOverviewLink: React.SFC<ErrorsOverviewLinkProps> = ({
```
nit: what do you think about `RelatedErrorsBadge` or `ErrorCountBadge` or something else that more describes what the component is, and not where it links to.

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/WaterfallItem.tsx:
```diff
@@ -154,12 +167,20 @@ export function WaterfallItem({
+      }}
+      onMouseLeave={() => {
+        setIsHovering(false);
+      }}
```
Doing this with js could be a little sluggish, since the entire component is re-rendered on every hover. What about some good ol' css?

```less
.parent-container {
  .child-container {
    display: none;
  }

  &:hover .child-container {
    display: block;
  }
}
```

sqren replied:
Btw. I noticed that it feels pretty snappy - so if the css approach it not doable it's probably fin as-is.

ogupte committed:
[APM] Renamed ErrorsOverviewLink -> ErrorCountBadge. Differentiate the labeled
vs non-labeled variaions of ErrorCountBadge. Replace state hooks for hovered
styles with plain css :hover

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5957/)

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/WaterfallItem.tsx:
```diff
@@ -41,8 +46,13 @@ const Container = styled<IContainerStyleProps, 'div'>('div')`
     background-color: ${theme.euiColorLightestShade};
+
+    ${DurationStyled} {
+      display: block;
```
You talked about using `visibility` instead of `display` to avoid the text to "jump". Was that not needed?

ogupte committed:
[APM] Revert duration display on hover to render display all the time.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6241/)

sqren commented on x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/WaterfallItem.tsx:
```diff
@@ -105,9 +108,9 @@ function PrefixIcon({ item }: { item: IWaterfallItem }) {
 }
 
-function Duration({ item }: { item: IWaterfallItem }) {
+function Duration({ item, ...props }: { item: IWaterfallItem }) {
```
Is `...props` still needed?

ogupte replied:
fixed

ogupte committed:
remove unnecessary props object spread

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6289/)
